pkg/operator/controllers/machineset/machineset_controller.go

```
@@ -0,0 +1,99 @@
+func NewReconciler(log *logrus.Entry, arocli aroclient.Interface, maocli maoclient.Interface) *Reconciler {
+	return &Reconciler{
+		log:    log,
+		arocli: arocli,
+		maocli: maocli,
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	instance, err := r.arocli.AroV1alpha1().Clusters().Get(ctx, arov1alpha1.SingletonClusterName, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if !instance.Spec.Features.ReconcileMachineSet {
+		return reconcile.Result{}, nil
+	}
+
+	machinesets, err := r.maocli.MachineV1beta1().MachineSets(machineSetsNamespace).List(ctx, metav1.ListOptions{LabelSelector: "machine.openshift.io/cluster-api-machine-role=worker"})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	aroMachineset, err := r.maocli.MachineV1beta1().MachineSets(machineSetsNamespace).Get(ctx, request.Name, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Count amount of total current worker replicas (including custom MachineSets, if present)
+	replicaCount := 0
+	for _, machineset := range machinesets.Items {
+		if machineset.Spec.Replicas != nil {
+			replicaCount += int(*machineset.Spec.Replicas)
+		}
+	}
+
+	// Verify that the MachineSet that was modified matches the cluster's InfraID, otherwise ignore it
+	matches, err := regexp.Match(instance.Spec.InfraID, []byte(aroMachineset.ObjectMeta.Name))
+	if err != nil {
+		r.log.Error(err)
+	}
+
+	if replicaCount < 3 && matches {
+		r.log.Info("Found less than 3 worker replicas. The MachineSet controller will attempt scaling.")
+		aroMachineset.Spec.Replicas = to.Int32Ptr(int32(3-replicaCount) + *aroMachineset.Spec.Replicas) // Add replicas to the object, and call Update
+		_, err := r.maocli.MachineV1beta1().MachineSets(machineSetsNamespace).Update(ctx, aroMachineset, metav1.UpdateOptions{})
```

If there's an error inside this block, it will not be returned now. It will return `err` from outside this scope, last set on line 73.
```suggestion
		_, err = r.maocli.MachineV1beta1().MachineSets(machineSetsNamespace).Update(ctx, aroMachineset, metav1.UpdateOptions{})
```

---

I would prefer having `return err` inside of `if`s instead of re-using `err` and returning it once at the end.